wallet/src/de/schildbach/wallet/service/CoinJoinService.kt

```diff
@@ -163,34 +164,45 @@ class CoinJoinMixingService @Inject constructor(
     }
 
     private suspend fun updateState(mixingStatus: MixingStatus, networkStatus: NetworkStatus, blockChain: AbstractBlockChain?) {
-        setBlockchain(blockChain)
-
-        when {
-            networkStatus == NetworkStatus.UNKNOWN -> return
-            mixingStatus == MixingStatus.MIXING && networkStatus == NetworkStatus.CONNECTED && isBlockChainSet -> {
-                // start mixing
-                prepareMixing()
-                startMixing()
-            }
-            this.mixingStatus == MixingStatus.MIXING && mixingStatus == MixingStatus.FINISHED -> {
-                // finish mixing
-                stopMixing()
-                setMixingComplete()
-            }
-            mixingStatus == MixingStatus.MIXING && this.networkStatus == NetworkStatus.CONNECTED && networkStatus == NetworkStatus.NOT_AVAILABLE -> {
-                // pause mixing
-                stopMixing()
-            }
-            mixingStatus == MixingStatus.PAUSED && this.networkStatus == NetworkStatus.CONNECTING && networkStatus == NetworkStatus.CONNECTED && isBlockChainSet -> {
-                // resume mixing
-                prepareMixing()
-                startMixing()
+        updateMutex.lock()
+        log.info("coinjoin-updateState: $mixingStatus, $networkStatus, ${blockChain != null}")
+        try {
+            setBlockchain(blockChain)
+            log.info("coinjoin-updateState: $mixingStatus, $networkStatus, ${blockChain != null}")
+            this.networkStatus = networkStatus
+            this.mixingStatus = mixingStatus
```

Comment:
For some reason, these lines were not executed if they were at the end of the function if mixingStatus = **MIXING** and if the app was restarted after being force closed.